addons/web/controllers/json.py

```diff
@@ -0,0 +1,254 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import ast
+import functools
+import logging
+from collections import defaultdict
+from datetime import date
+from dateutil.relativedelta import relativedelta
+from http import HTTPStatus
+from lxml import etree
+from urllib.parse import urlencode
+
+from werkzeug.exceptions import BadRequest
+
+from odoo import http
+from odoo.exceptions import AccessError
+from odoo.http import request
+from odoo.models import regex_object_name
+from odoo.osv import expression
+from odoo.tools import lazy
+from odoo.tools.safe_eval import safe_eval
+from .utils import get_action_triples
+
+
+_logger = logging.getLogger(__name__)
+
+
+class WebJsonController(http.Controller):
+
+    @http.route('/json/<path:subpath>', auth='public', type='http', readonly=True)
+    def web_json(self, subpath, **kwargs):
+        return self.web_json_18_0(subpath, **kwargs, redirect=True)
+
+    @http.route('/json/18.0/<path:subpath>', auth='user', type='http', readonly=True)
+    def web_json_18_0(self, subpath, **kwargs):
+        env = request.env
+        if not request.env.user.has_group('base.group_allow_export'):
+            raise AccessError(env._("You need export permissions to use the /json route"))
+
+        # redirect when the kwargs have changed
+        param_list = set(kwargs)
+        kwargs.pop("redirect", None)  # pop redirect, so it is forced in check_redirect
+
+        def check_redirect():
+            if set(param_list) == set(kwargs):
+                return None
+            # for domains, make chars as safe
+            encoded_kwargs = urlencode(kwargs, safe="()[], '\"")
+            return request.redirect(
+                f'/json/18.0/{subpath}?{encoded_kwargs}',
+                HTTPStatus.TEMPORARY_REDIRECT
+            )
+
+        def get_action_triples_():
+            try:
+                yield from get_action_triples(env, subpath, start_pos=1)
+            except ValueError as exc:
+                raise BadRequest(exc.args[0])
+
+        # Hack for OXP. We are not sure yet if we wanna run all server
+        # actions, but we are sure we want to run those ones. TODO: find
+        # a better way to do it.
+        allowed_server_action_paths = {'crm'}
```

Comment:
Wow, we need to get rid of that one for the actual release...

Reply:
Not yet, I will have to check with @Julien00859 for how long it is needed.

Reply:
Seriously?

Reply:
It was added in a hurry for the community days. The set can be taken from a classmethod defined on the controller. That classmethod can be overriden inside crm to add 'crm'.

Reply:
For now, I'm only checking if the action has a `path`. I will try to add `usage=window_action` to clearly define the wanted result.

Reply:
@odony I have made sure that the DB connection is read-only.

Besides, I have added a commit with a new `usage=ir_window_action`. it seems to work. Previously I was executing `usage=ir_server_action and path`.
I'm not sure what is the preferred method that we should keep.